graph.py

```diff
@@ -0,0 +1,50 @@
+#!/usr/bin/python3
+# graph.py
+from collections import defaultdict
+
+
+class Graph():
+    def __init__(self, vertices):
+        self.graph = defaultdict(list)
+        self.verticies = vertices
+
+    def add_edge(self, u, v):
+        self.graph[u].append(v)
+
+    # internal
+    def _is_cyclic_util(self, start_vertex):
+        current_vertex = start_vertex + 0
+        visited = [current_vertex]
+        while len(self.graph[current_vertex]) > 0 and not(self.graph[current_vertex][0] == start_vertex):
+            current_vertex = self.graph[current_vertex][0]
+            visited.append(current_vertex + 0)
+        if len(self.graph[current_vertex]) > 0:
+            return [visited]
+        return []
+
+    # rotates cycle so it starts with smallest id
+    def normalize_cycle(self, a):
+        b = list(a)
+        b.sort()
+        loc = a.index(b[0])
+        g = [0] * len(a)
+        for x in range(0, len(a)):
+            g[x - loc] = a[x]
+        return g
+
+    # returns existing cycles in the form of lists of node ids which are in a cycle
+    def cycle_list(self):
+        cycles = []
+        for node in range(self.verticies):
```

Comment:
A minor optimization: you don't have to do it for every node. If you start with node A, and go to B, C, D, and then stop, you know that neither of A, B, C, D are in a cycle.